core-backend/common/src/utils.rs

```diff
@@ -0,0 +1,117 @@
+use alloc::string::String;
+
+pub(crate) fn smart_truncate(s: &mut String, max_bytes: usize) {
+    let mut last_byte = max_bytes;
+
+    if s.len() > last_byte {
+        while !s.is_char_boundary(last_byte) && last_byte != 0 {
+            last_byte = last_byte.saturating_sub(1);
+        }
+
+        s.truncate(last_byte);
+    }
```

Comment:
there is an easier alternative:
```rust
while s.len() > max_bytes {
    s.pop();
}
```

How do you look at it?

Reply:
what if s.len is 100k bytes?

Reply:
current solution has LAST_CHARACTER_BYTES_AMOUNT iterations in the worst case